management/networksecuritygroup/client.go

```diff
@@ -0,0 +1,235 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var securityGroups []*SecurityGroup
+	if err = xml.Unmarshal(response, &securityGroups); err != nil {
+		return nil, err
+	}
+
+	return securityGroups, nil
+}
+
+// AddNetworkSecurityToSubnet associates the network security group with
+// specified subnet in a virtual network
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn913822.aspx
+func (sg *SecurityGroupClient) AddNetworkSecurityToSubnet(
+	name string,
+	subnet string,
+	virtualNetwork string) (management.OperationId, error) {
+	if name == "" {
+		return "", fmt.Errorf(errParamNotSpecified, "name")
+	}
+	if subnet == "" {
+		return "", fmt.Errorf(errParamNotSpecified, "subnet")
+	}
+	if virtualNetwork == "" {
+		return "", fmt.Errorf(errParamNotSpecified, "virtualNetwork")
+	}
+
+	data, err := xml.Marshal(SecurityGroup{Name: name})
+	if err != nil {
+		return "", err
+	}
+
+	requestURL := fmt.Sprintf(addSecurityGroupToSubnetURL, virtualNetwork, subnet)
+	return sg.client.SendAzurePostRequest(requestURL, data)
+}
+
+// GetNetworkSecurityGroupForSubnet returns information about the network
+// security group associated with a subnet
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn913817.aspx
+func (sg *SecurityGroupClient) GetNetworkSecurityGroupForSubnet(
+	subnet string,
+	virtualNetwork string) (*SecurityGroup, error) {
+	if subnet == "" {
+		return nil, fmt.Errorf(errParamNotSpecified, "subnet")
+	}
+	if virtualNetwork == "" {
+		return nil, fmt.Errorf(errParamNotSpecified, "virtualNetwork")
+	}
+
+	requestURL := fmt.Sprintf(getSecurityGroupForSubnetURL, virtualNetwork, subnet)
+	response, err := sg.client.SendAzureGetRequest(requestURL)
+	if err != nil {
+		return nil, err
+	}
+
+	var securityGroup SecurityGroup
+	if err := xml.Unmarshal(response, &securityGroup); err != nil {
+		return nil, err
+	}
+
+	return &securityGroup, nil
+}
+
+// RemoveNetworkSecurityGroupFromSubnet removes the association of the
+// specified network security group from the specified subnet
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn913820.aspx
+func (sg *SecurityGroupClient) RemoveNetworkSecurityGroupFromSubnet(
+	name string,
+	subnet string,
+	virtualNetwork string) (management.OperationId, error) {
+	if name == "" {
+		return "", fmt.Errorf(errParamNotSpecified, "name")
+	}
+	if subnet == "" {
+		return "", fmt.Errorf(errParamNotSpecified, "subnet")
+	}
+	if virtualNetwork == "" {
+		return "", fmt.Errorf(errParamNotSpecified, "virtualNetwork")
+	}
+
+	requestURL := fmt.Sprintf(removeSecurityGroupFromSubnetURL, virtualNetwork, subnet, name)
+	return sg.client.SendAzureDeleteRequest(requestURL)
+}
+
+// SetNetworkSecurityGroupRule adds or updates a network security rule that
+// is associated with the specified network security group
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn913819.aspx
+func (sg *SecurityGroupClient) SetNetworkSecurityGroupRule(
+	securityGroup string,
+	rule *Rule) (management.OperationId, error) {
```

Comment:
I don't think `rule` should be a pointer here (nil is not really a valid input, so let language's type system handle that for you).